File: src/weaponinventorymod/gui/StockReviewFilterListModel.java
package weaponinventorymod.gui;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

final class StockReviewFilterListModel {
    private StockReviewFilterListModel() {
    }

    static List<WimGuiListRow<StockReviewAction>> build(StockReviewState state) {
        List<WimGuiListRow<StockReviewAction>> rows = new ArrayList<WimGuiListRow<StockReviewAction>>();
        Set<StockReviewFilter> active = state.getActiveFilters();
        if (!active.isEmpty()) {
            for (StockReviewFilter filter : StockReviewFilter.values()) {
                if (active.contains(filter)) {
                    rows.add(StockReviewListRow.filter(filter.getLabel(), true,
                            StockReviewAction.toggleFilter(filter), false));
                }
            }
        }
        for (StockReviewFilterGroup group : StockReviewFilterGroup.values()) {
            addGroup(rows, state, group, !active.isEmpty() || group.ordinal() > 0);
        }
        return rows;
    }

    private static void addGroup(List<WimGuiListRow<StockReviewAction>> rows,
                                 StockReviewState state,
                                 StockReviewFilterGroup group,
                                 boolean topGap) {
        boolean expanded = state.isExpanded(group);
        Set<StockReviewFilter> activeInGroup = StockReviewFilters.activeInGroup(state.getActiveFilters(), group);
        String label = WimGuiToggleHeading.countedLabel(group.getLabel(), activeInGroup.size(), expanded);
        rows.add(StockReviewListRow.filterHeading(label, StockReviewAction.toggle(group), topGap));
        if (!expanded) {
            return;
        }
        for (StockReviewFilter filter : StockReviewFilter.values()) {
            if (!group.equals(filter.getGroup())) {
                continue;
            }
            boolean active = state.isFilterActive(filter);
            if (active) {
                continue;
            }
            rows.add(StockReviewListRow.filter(filter.getLabel(), false,
                    StockReviewAction.toggleFilter(filter), false));
        }
    }
}

File: src/main/kotlin/weaponsprocurement/ui/stockreview/state/StockReviewFilterListModel.kt
package weaponsprocurement.ui.stockreview.state

import weaponsprocurement.ui.stockreview.actions.*
import weaponsprocurement.ui.stockreview.state.*
import weaponsprocurement.ui.stockreview.rows.*
import weaponsprocurement.ui.stockreview.tooltips.*
import weaponsprocurement.ui.stockreview.rendering.*
import weaponsprocurement.ui.stockreview.trade.*

import weaponsprocurement.ui.*

import java.util.ArrayList

class StockReviewFilterListModel private constructor() {
    companion object {
        @JvmStatic
        fun build(state: StockReviewState): List<WimGuiListRow<StockReviewAction>> {
            val rows = ArrayList<WimGuiListRow<StockReviewAction>>()
            val active = state.getActiveFilters()
            if (active.isNotEmpty()) {
                for (filter in StockReviewFilter.values()) {
                    if (active.contains(filter)) {
                        rows.add(
                            StockReviewListRow.filter(
                                filter.label,
                                true,
                                StockReviewAction.toggleFilter(filter),
                                false,
                                StockReviewTooltips.filter(filter, true),
                            ),
                        )
                    }
                }
            }
            for (group in StockReviewFilterGroup.values()) {
                addGroup(rows, state, group, active.isNotEmpty() || group.ordinal > 0)
            }
            return rows
        }

        private fun addGroup(
            rows: MutableList<WimGuiListRow<StockReviewAction>>,
            state: StockReviewState,
            group: StockReviewFilterGroup,
            topGap: Boolean,
        ) {
            val expanded = state.isExpanded(group)
            val activeInGroup = StockReviewFilters.activeInGroup(state.getActiveFilters(), group)
            val label = WimGuiToggleHeading.countedLabel(group.label, activeInGroup.size, expanded)
            rows.add(
                StockReviewListRow.filterHeading(
                    label,
                    StockReviewAction.toggle(group),
                    topGap,
                    StockReviewTooltips.filterHeading(group),
                ),
            )
            if (!expanded) {
                return
            }
            for (filter in StockReviewFilter.values()) {
                if (group != filter.group) {
                    continue
                }
                val active = state.isFilterActive(filter)
                if (active) {
                    continue
                }
                rows.add(
                    StockReviewListRow.filter(
                        filter.label,
                        false,
                        StockReviewAction.toggleFilter(filter),
                        false,
                        StockReviewTooltips.filter(filter, false),
                    ),
                )
            }
        }
    }
}